File: figures/05-dynamic-programming/snippet-max-subarray-kadane.c
static bool max_subarray_kadane(const long long *a,size_t n,MaxSubarrayResult *out){
    if(!a||!out||n==0)return false;
    MaxSubarrayResult ending={a[0],0,0},best=ending;
    for(size_t i=1;i<n;++i){
        MaxSubarrayResult extend={ending.sum+a[i],ending.start,i},restart={a[i],i,i};
        ending=better(restart,extend)?restart:extend;if(better(ending,best))best=ending;
    }*out=best;return true;
}

File: figures/05-dynamic-programming/snippet-max-subarray-kadane.py
def max_subarray_kadane(a):
    ending = (a[0], 0, 0)
    best = ending
    for i in range(1, len(a)):
        extend = (ending[0] + a[i], ending[1], i)
        restart = (a[i], i, i)
        ending = restart if _better(restart, extend) else extend
        if _better(ending, best):
            best = ending
    return best
